File: macos/LocalCloud/Engine/Mesh.swift
import Foundation
import Observation

/// Everything the app knows about the mesh, and every way it can change it.
///
/// Lives on the main actor and never calls the engine directly: every call goes
/// through [EngineWorker], which is neither the main thread nor two threads at
/// once. What comes back is a whole [MeshState], replaced in one assignment, so
/// a view never sees half of an update.
///
/// Isolated to the main actor by hand, not by the project default. The default
/// would isolate the generated bindings too, and the engine's contract is the
/// opposite of that - see the note in `EngineWorker`.
@Observable
@MainActor
final class Mesh {

    private(set) var state = MeshState()
    private(set) var transfers: [String: Transfer] = [:]
    private(set) var notice: Notice?

    @ObservationIgnored private var worker: EngineWorker?
    @ObservationIgnored private var begun = false
    @ObservationIgnored private var tasks: [Task<Void, Never>] = []
    @ObservationIgnored private var noticeCount = 0

    /// Events, as they leave the engine's own thread.
    ///
    /// The engine's contract is that a listener must not block and must not
    /// call back in and wait. Yielding to a stream does neither. The buffer is
    /// the same 256 the engine keeps on its side, and for the same reason: a
    /// backlog that is capped can say it overflowed, where one that grows just
    /// grows.
    @ObservationIgnored private let events: AsyncStream<EngineEvent>
    @ObservationIgnored private let deliver: @Sendable (EngineEvent) -> Void

    /// Conflated on purpose. A burst of thirty block-progress events should
    /// cause one re-read of the catalog, not thirty.
    @ObservationIgnored private let refreshRequests: AsyncStream<Void>
    @ObservationIgnored private let requestRefresh: @Sendable () -> Void

    init() {
        let (events, eventsIn) = AsyncStream<EngineEvent>.makeStream(
            bufferingPolicy: .bufferingNewest(256)
        )
        self.events = events
        deliver = { eventsIn.yield($0) }

        let (refreshes, refreshesIn) = AsyncStream<Void>.makeStream(
            bufferingPolicy: .bufferingNewest(1)
        )
        refreshRequests = refreshes
        requestRefresh = { refreshesIn.yield(()) }
    }

    // -- Lifecycle ------------------------------------------------------------

    /// Constructs the engine, registers for events, and starts it.
    ///
    /// The listener is registered before the engine is started, which is what
    /// the engine's own backlog exists to make unnecessary - but arriving late
    /// to your own events is still a poor way to run a window.
    func begin() {
        guard !begun else { return }
        begun = true

        Task { [self] in
            let worker: EngineWorker
            do {
                worker = try EngineWorker()
            } catch {
                state.ready = true
                state.fatal = error.sentenceForUser
                return
            }
            self.worker = worker

            let bridge = EventBridge(deliver)
            await worker.setEventListener(bridge)

            tasks.append(Task { [self] in
                for await event in events { handle(event) }
            })
            tasks.append(Task { [self] in
                for await _ in refreshRequests { await refresh() }
            })

            // Discovery and local changes both announce themselves, but a
            // catalog that changed on a peer does not: replication is pull-only,
            // and the engine re-reads its peers every 30 seconds without saying
            // so. Nor does a device going quiet produce an event. So the window
            // is kept honest by asking, cheaply, on a timer.
            tasks.append(Task { [self] in
                while !Task.isCancelled {
                    try? await Task.sleep(for: .seconds(2))
                    if state.running { requestRefresh() }
                }
            })

            do {
                try await worker.start()
            } catch {
                state.ready = true
                state.fatal = error.sentenceForUser
            }
            requestRefresh()
        }
    }

    /// Stops the engine, on the way out.
    ///
    /// Awaited rather than fired off, because the process is about to end and a
    /// mDNS goodbye that never got sent leaves this device advertised on the
    /// network until every peer times it out.
    func end() async {
        for task in tasks { task.cancel() }
        tasks = []
        await worker?.stop()
    }

    // -- Reading --------------------------------------------------------------

    private func refresh() async {
        guard let worker else { return }
        state = await worker.snapshot()
    }

    // -- Doing ----------------------------------------------------------------

    func importFiles(_ urls: [URL]) {
        Task { [self] in
            for url in urls {
                // A file chosen in an open panel, or dropped on the window, is
                // outside the sandbox until this is called. The engine reads it
                // with plain POSIX calls and has no idea any of this is going
                // on, which is the point: the app owns the sandbox, the engine
                // owns the file.
                let scoped = url.startAccessingSecurityScopedResource()
                defer { if scoped { url.stopAccessingSecurityScopedResource() } }

                await attempt {
                    try await self.worker?.importFile(
                        at: url.path(percentEncoded: false),
                        named: url.lastPathComponent
                    )
                }
            }
        }
    }

    func share(_ item: Item, to deviceIds: [String]) {
        act { try await $0.share(item.id, to: deviceIds) }
    }

    func pull(_ item: Item) {
        act { try await $0.pull(item.id) }
    }

    func deleteHere(_ item: Item) {
        Task { [self] in
            guard let worker else { return }
            let outcome = await attempt { try await worker.deleteHere(item.id) }
            guard let outcome else { return }
            announce(
                outcome.trashed
                    ? "That was the last copy. It is in the trash for 30 days."
                    : "Deleted here. \(copies(outcome.remainingCopies)) left elsewhere."
            )
        }
    }

    func delete(_ item: Item, from deviceId: String) {
        act { try await $0.delete(item.id, from: deviceId) }
    }

    func beginPairing(with device: DiscoveredDevice) async -> String? {
        guard let worker else { return nil }
        return await attempt { try await worker.beginPairing(with: device.deviceId) }
    }

    func confirmPairing(with deviceId: String, code: String) async -> Bool {
        guard let worker else { return false }
        return await attempt { try await worker.confirmPairing(with: deviceId, code: code) } != nil
    }

    func cancelPairing() {
        act { await $0.cancelPairing() }
    }

    func unpair(_ device: PairedDevice) {
        act { try await $0.unpair(device.id) }
    }

    func rename(to name: String) {
        act { try await $0.rename(to: name) }
    }

    func restore(_ item: TrashItem) {
        act { try await $0.restore(item.id) }
    }

    func destroy(_ item: TrashItem) {
        act { try await $0.destroy(item.id) }
    }

    func resolve(_ collision: PendingCollision, as resolution: CollisionResolution) {
        act { try await $0.resolve(collision.id, as: resolution) }
    }

    /// One place where an engine call is made, fails, and is reported.
    ///
    /// Returning nil rather than throwing, because every caller is a click:
    /// there is nothing to recover, only something to say. The refresh
    /// afterwards is unconditional - a call that failed part way through may
    /// still have changed something.
    @discardableResult
    private func attempt<T>(_ body: () async throws -> T) async -> T? {
        defer { requestRefresh() }
        do {
            return try await body()
        } catch {
            announce(error.sentenceForUser, .failure)
            return nil
        }
    }

    /// The common shape: do one thing to the engine, say so if it refuses.
    private func act(_ body: @escaping (EngineWorker) async throws -> Void) {
        Task { [self] in
            guard let worker else { return }
            await attempt { try await body(worker) }
        }
    }

    // -- Events ---------------------------------------------------------------

    private func handle(_ event: EngineEvent) {
        switch event {
        case let .sendProgress(fileId, deviceId, done, total):
            put(Transfer(
                fileId: fileId, direction: .sending, peerId: deviceId,
                blocksDone: done, blocksTotal: total
            ))

        case let .receiveProgress(fileId, done, total):
            put(Transfer(
                fileId: fileId, direction: .receiving, peerId: nil,
                blocksDone: done, blocksTotal: total
            ))

        case let .fileSent(fileId, path, deviceId):
            clear(fileId, deviceId)
            announce("\(displayName(path)) reached \(nameOf(deviceId)).")

        case let .shareFailed(fileId, path, deviceId, reason):
            clear(fileId, deviceId)
            announce(
                "\(displayName(path)) did not reach \(nameOf(deviceId)): \(reason)",
                .failure
            )

        case let .fileDownloaded(fileId, path):
            clear(fileId, nil)
            announce("\(displayName(path)) is now on this device.")

        case let .pullFailed(fileId, reason):
            clear(fileId, nil)
            announce("Could not take a copy: \(reason)", .failure)

        case let .devicePaired(_, name):
            announce("Paired with \(name).")

        case let .pairingFailed(_, reason):
            announce("Pairing failed: \(reason)", .failure)

        case let .nameCollision(requestedPath, keptAs):
            announce(
                "\(displayName(requestedPath)) was already taken, so it arrived as \(displayName(keptAs))."
            )

        case let .deleteRequestDeferred(_, deviceId, _):
            announce(
                "\(nameOf(deviceId)) could not be reached, so the delete will travel with the catalog."
            )

        case let .engineFailed(reason):
            announce(reason, .failure)

        // Everything else is a change to state the refresh below picks up.
        // There is nothing to say about it that the window will not show.
        default:
            break
        }
        requestRefresh()
    }

    private func put(_ transfer: Transfer) {
        transfers[Transfer.key(transfer.fileId, transfer.peerId)] = transfer
    }

    private func clear(_ fileId: String, _ peerId: String?) {
        transfers[Transfer.key(fileId, peerId)] = nil
    }

    private func nameOf(_ deviceId: String) -> String {
        state.paired.first { $0.id == deviceId }?.name
            ?? state.visible.first { $0.deviceId == deviceId }?.name
            ?? shortId(deviceId)
    }

    // -- Saying ---------------------------------------------------------------

    func announce(_ text: String, _ kind: Notice.Kind = .info) {
        noticeCount += 1
        let notice = Notice(id: noticeCount, text: text, kind: kind)
        self.notice = notice

        // Cleared by id rather than unconditionally, so a second notice arriving
        // in the meantime is not wiped by the first one's timer.
        Task { [self] in
            try? await Task.sleep(for: .seconds(8))
            if self.notice?.id == notice.id { self.notice = nil }
        }
    }

    func dismissNotice() {
        notice = nil
    }

    private func copies(_ count: Int64) -> String {
        count == 1 ? "1 copy" : "\(count) copies"
    }
}

/// Carries events from the engine's thread to the stream the app reads.
///
/// Its own type, because the engine calls it from a thread of its own and knows
/// nothing about actors. It holds a closure and no state, so there is nothing
/// here for two threads to disagree about.
private final class EventBridge: EventListener {
    private let deliver: @Sendable (EngineEvent) -> Void

    init(_ deliver: @escaping @Sendable (EngineEvent) -> Void) {
        self.deliver = deliver
    }

    func onEvent(event: EngineEvent) {
        deliver(event)
    }
}

File: web/ui/src/mesh.ts
import { createSignal } from 'solid-js'
import { createStore, produce, reconcile } from 'solid-js/store'
import type { Collision, Item, Notice, Progress, Snapshot, TrashItem } from './types'

/**
 * Everything the page knows, and every way it can change it.
 *
 * One module-level store rather than a context. There is one engine, one
 * connection to it, and one page; a provider would be ceremony around a
 * singleton.
 *
 * The whole state arrives as one object and is applied with `reconcile`, which
 * is the point of doing the join on the server. Solid then walks the two trees
 * and touches only the signals that actually differ, so a snapshot arriving
 * every two seconds with one changed byte updates one text node — not a list.
 */

const empty: Snapshot = {
  device: { id: '', shortId: '', name: '', platform: '', running: false, syncDir: '' },
  items: [],
  trash: [],
  visible: [],
  paired: [],
  offers: [],
  collisions: [],
  deferredDeletes: 0,
}

const [state, setState] = createStore<Snapshot>(empty)

/**
 * Kept apart from the snapshot on purpose.
 *
 * Block progress arrives up to a thousand times for one large file. Putting it
 * in the snapshot would mean re-reading the catalog for each one; keeping it
 * here means a transfer bar updates and nothing else on the page is touched.
 */
const [transfers, setTransfers] = createStore<Record<string, Progress>>({})

const [notice, setNotice] = createSignal<Notice | null>(null)
const [connected, setConnected] = createSignal(false)

let noticeCount = 0
let noticeTimer: ReturnType<typeof setTimeout> | undefined

export { state, transfers, notice, connected }

export const key = (fileId: string, deviceId: string | null) => `${fileId}@${deviceId ?? 'here'}`

export function say(text: string, failure = false) {
  noticeCount += 1
  setNotice({ text, failure, id: noticeCount })
  clearTimeout(noticeTimer)
  noticeTimer = setTimeout(() => setNotice(null), failure ? 9000 : 5000)
}

export const dismiss = () => setNotice(null)

// -- The connection ----------------------------------------------------------

/**
 * One EventSource for the life of the page.
 *
 * The browser reconnects it on its own, so there is no retry loop here — only
 * a flag, because a page that is drawing a two-minute-old catalog should say
 * so rather than look correct.
 */
export function connect() {
  const source = new EventSource('/api/events')

  source.addEventListener('open', () => setConnected(true))
  source.addEventListener('error', () => setConnected(false))

  source.addEventListener('state', (event) => {
    setConnected(true)
    setState(reconcile(JSON.parse((event as MessageEvent).data) as Snapshot, { key: 'id' }))
  })

  source.addEventListener('progress', (event) => {
    const progress = JSON.parse((event as MessageEvent).data) as Progress
    setTransfers(key(progress.fileId, progress.deviceId), progress)
  })

  source.addEventListener('done', (event) => {
    const { fileId, deviceId } = JSON.parse((event as MessageEvent).data)
    setTransfers(produce((all) => { delete all[key(fileId, deviceId ?? null)] }))
  })

  source.addEventListener('notice', (event) => {
    const { text, failure } = JSON.parse((event as MessageEvent).data)
    say(text, failure)
  })
}

// -- Asking it to do things --------------------------------------------------

/**
 * One place where a request is made, fails, and is reported.
 *
 * Returns null rather than throwing, because every caller is a click: there is
 * nothing to recover, only something to say. The engine writes a sentence for
 * every one of its failures and this side does not improve on it.
 */
async function post<T = unknown>(path: string, body?: unknown): Promise<T | null> {
  try {
    const response = await fetch(path, {
      method: 'POST',
      headers: body ? { 'content-type': 'application/json' } : undefined,
      body: body ? JSON.stringify(body) : undefined,
    })
    const payload = await response.json().catch(() => null)
    if (!response.ok) {
      say(payload?.error ?? `That did not work (${response.status}).`, true)
      return null
    }
    return payload as T
  } catch {
    say('The engine is not answering. Is it still running?', true)
    return null
  }
}

/**
 * Sends the file as the request body rather than as a form.
 *
 * A `File` is a stream, so nothing here holds the bytes: the browser reads it
 * from disk as the socket drains and the server writes it out the other side.
 * A multipart form would have both ends buffering a copy for no reason.
 */
export async function importFiles(files: FileList | File[]) {
  for (const file of Array.from(files)) {
    try {
      const response = await fetch(`/api/import?name=${encodeURIComponent(file.name)}`, {
        method: 'POST',
        body: file,
      })
      if (!response.ok) {
        const payload = await response.json().catch(() => null)
        say(payload?.error ?? `${file.name} could not be added.`, true)
      }
    } catch {
      say(`${file.name} could not be added.`, true)
    }
  }
}

export const share = (item: Item, deviceIds: string[]) =>
  post('/api/share', { fileId: item.id, deviceIds })

export const pull = (item: Item) => post('/api/pull', { fileId: item.id })

export async function deleteHere(item: Item) {
  const outcome = await post<{ remainingCopies: number; trashed: boolean }>(
    '/api/delete-here',
    { fileId: item.id },
  )
  if (!outcome) return
  say(
    outcome.trashed
      ? 'That was the last copy. It is in the trash for 30 days.'
      : `Deleted here. ${outcome.remainingCopies === 1 ? '1 copy' : `${outcome.remainingCopies} copies`} left elsewhere.`,
  )
}

export const deleteFrom = (item: Item, deviceId: string) =>
  post('/api/delete-copy', { fileId: item.id, deviceId })

export const startPairing = (deviceId: string) =>
  post<{ code: string }>('/api/pair/start', { deviceId })

export const confirmPairing = (deviceId: string, code: string) =>
  post('/api/pair/confirm', { deviceId, code })

export const cancelPairing = () => post('/api/pair/cancel')

export const unpair = (deviceId: string) => post('/api/unpair', { deviceId })

export const rename = (name: string) => post('/api/rename', { name })

export const restore = (item: TrashItem) => post('/api/trash/restore', { fileId: item.id })

export const destroy = (item: TrashItem) => post('/api/trash/destroy', { fileId: item.id })

export const resolveCollision = (collision: Collision, keepBoth: boolean) =>
  post('/api/collision', { collisionId: collision.id, keepBoth })
